src/CMakeLists.txt

```diff
@@ -127,8 +138,26 @@ add_library(caideInliner STATIC
 
 target_include_directories(caideInliner SYSTEM PRIVATE ${CLANG_INCLUDE_DIRS} ${LLVM_INCLUDE_DIRS})
 target_compile_definitions(caideInliner PRIVATE ${CLANG_DEFINITIONS} ${LLVM_DEFINITIONS})
-target_link_libraries(caideInliner PRIVATE
-    clangAST clangBasic clangFrontend clangLex clangRewrite clangSema clangTooling)
+
+if(CAIDE_INLINER_LINK_CLANG_DYLIB)
+    set(CAIDE_INLINER_CLANG_LIBS clang-cpp)
+else(CAIDE_INLINER_LINK_CLANG_DYLIB)
+    set(CAIDE_INLINER_CLANG_LIBS clangAST
+                                 clangBasic
+                                 clangFrontend
+                                 clangLex
+                                 clangRewrite
+                                 clangSema
+                                 clangTooling)
+endif(CAIDE_INLINER_LINK_CLANG_DYLIB)
+
+if(CAIDE_INLINER_LINK_LLVM_DYLIB)
+    set(CAIDE_INLINER_LLVM_LIBS  LLVM)
```

Comment:
Do we actually need this? We don't depend on LLVM directly, so it should be brought up automatically by clang dependencies.

Reply:
Currently, the automatic resolving of dependencies in the case of `CAIDE_INLINER_LINK_LLVM_DYLIB=ON` and `CAIDE_USE_SYSTEM_CLANG=ON` does not seem to work. I will investigate the problem in more detail. It would be nice if the LLVM dependencies could be resolved automatically.

Reply:
It is definitely possible to automatically determine the dependencies of `clang-cpp` in the case of `CLANG_LINK_CLANG_DYLIB=ON`, since clang is shipped with the following CMake configuration fragment:
```
# Import target "clang-cpp" for configuration "Release"
set_property(TARGET clang-cpp APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
set_target_properties(clang-cpp PROPERTIES
  IMPORTED_LINK_DEPENDENT_LIBRARIES_RELEASE "LLVM"
  IMPORTED_LOCATION_RELEASE "${_IMPORT_PREFIX}/lib/libclang-cpp.so.17"
  IMPORTED_SONAME_RELEASE "libclang-cpp.so.17"
)
```

The necessary dependency can now be obtained by reading the target property `IMPORTED_LINK_DEPENDENT_LIBRARIES_<CONFIG>` where `<CONFIG>` is one of the configurations from `IMPORTED_CONFIGURATIONS`. I have already implemented this for testing and I can report that the dependency resolution works. However, linking still fails because these dependencies are not added to the `caideInliner` target, but are just dumped into a file.

Why is this automatic dependency resolution part of the CMake project if it has no influence on the linker process?

How should we proceed here? Theoretically, what is missing here is adding the resolved dependencies to the target so that the linker knows all the necessary directly and indirectly dependent dependencies.

Reply:
> linking still fails because these dependencies are not added to the caideInliner target, but are just dumped into a file.

If you mean caide-libs.txt file -- that's just for https://github.com/slycelote/caide project, which is not built by CMake and hence needs an "out-of-band" information about how to link the final executable... Linking within CMake shouldn't need explicit transitive dependencies -- and indeed it doesn't when we build only static clang/llvm libraries (with `CAIDE_USE_SYSTEM_CLANG=OFF`). Perhaps I misunderstand how cmake works with shared libraries...

Anyway, it's ok to keep this for now. Thanks for trying to make sense of it!

Reply:
Yeah, exactly, I mean the `caide-libs.txt` file . Thanks for the explanation and merging the changes.

Linking Clang is always easier with static libraries. Unfortunately, there are distributions (e.g. Archlinux) where Clang/LLVM is configured and built with `CLANG_LINK_CLANG_DYLIB=ON` and `LLVM_LINK_LLVM_DYLIB=ON`. This means that the artifacts are only shipped as shared libraries.

In the future, we could alternatively use the CMake utilities from Clang/LLVM, which directly manage the linker dependencies and support in- and out-of-tree builds with static and dynamic libraries. A use of the CMake utilities is presented in the LLVM repository in a [tool template for Clang-based tools](https://github.com/llvm/llvm-project/blob/04a6e0f1634f9a53120c27a30250d26dff4ada1c/clang-tools-extra/tool-template/CMakeLists.txt).